includes/amp-helper-functions.php

```diff
@@ -167,13 +167,133 @@ function amp_get_asset_url( $file ) {
 		. '<noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>';
 }
 
+/**
+ * Register default scripts for AMP components.
+ *
+ * @param WP_Scripts $wp_scripts Scripts.
+ */
+function amp_register_default_scripts( $wp_scripts ) {
+
+	// AMP Runtime.
+	$handle = 'amp-runtime';
+	$wp_scripts->add(
+		$handle,
+		'https://cdn.ampproject.org/v0.js',
+		array(),
+		null
+	);
+	$wp_scripts->add_data( $handle, 'amp_script_attributes', array(
+		'async' => true,
+	) );
+
+	// Shadow AMP API.
+	$handle = 'amp-shadow';
+	$wp_scripts->add(
+		$handle,
+		'https://cdn.ampproject.org/shadow-v0.js',
+		array(),
+		null
+	);
+	$wp_scripts->add_data( $handle, 'amp_script_attributes', array(
+		'async' => true,
+	) );
+
+	// Get all AMP components as defined in the spec.
+	$extensions = array();
+	foreach ( AMP_Allowed_Tags_Generated::get_allowed_tags() as $allowed_tag ) {
+		foreach ( $allowed_tag as $rule_spec ) {
+			if ( ! isset( $rule_spec[ AMP_Rule_Spec::TAG_SPEC ] ) ) {
+				continue;
+			}
+			$tag_spec = $rule_spec[ AMP_Rule_Spec::TAG_SPEC ];
+			if ( ! empty( $tag_spec['also_requires_tag_warning'] ) ) {
+				$extensions[] = strtok( $tag_spec['also_requires_tag_warning'][0], ' ' );
+			}
+			if ( ! empty( $tag_spec['requires_extension'] ) ) {
+				$extensions = array_merge( $extensions, $tag_spec['requires_extension'] );
+			}
+		}
+	}
+
+	/**
+	 * List of components that are custom elements.
+	 *
+	 * Per the spec, "Most extensions are custom-elements." In fact, there is only one custom template.
+	 *
+	 * @link https://github.com/ampproject/amphtml/blob/cd685d4e62153557519553ffa2183aedf8c93d62/validator/validator.proto#L326-L328
+	 * @link https://github.com/ampproject/amphtml/blob/cd685d4e62153557519553ffa2183aedf8c93d62/extensions/amp-mustache/validator-amp-mustache.protoascii#L27
+	 */
+	$custom_templates = array( 'amp-mustache' );
+
+	foreach ( $extensions as $extension ) {
+		$src = sprintf(
+			'https://cdn.ampproject.org/v0/%s-%s.js',
+			$extension,
+			'latest'
+		);
+
+		$wp_scripts->add(
+			$extension,
+			$src,
+			array( 'amp-runtime' ),
+			null
+		);
+		$attributes = array(
+			'async' => true,
+		);
+		if ( in_array( $extension, $custom_templates, true ) ) {
+			$attributes['custom-template'] = $extension;
+		} else {
+			$attributes['custom-element'] = $extension;
+		}
+		$wp_scripts->add_data( $extension, 'amp_script_attributes', $attributes );
```

Comment:
This really isn't necessary. In the script_loader_tag filter we can just look to see if the script comes from the AMP CDN and if so then automatically add async and the required custom attribute.